---
title: "Run EBPMF on Montoro et al data"
author: "DongyueXie"
date: "2023-03-17"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---



# Montoro et al data, droplet, 7193 cells

```{r,eval=F}
rare_idx = which(samples$tissue %in% c("Goblet","Ionocyte", "Neuroendocrine","Tuft"))
count_rare = counts[rare_idx,]
# remove genes that have no expression
count_rare = count_rare[,colSums(count_rare!=0)>0]
```

## Droplet plots 

We plot K = 7

```{r}
load('/project2/mstephens/pcarbo/git/single-cell-topics/data/droplet.RData')
source("~/Rpackages/gsmash/code/poisson_STM/structure_plot.R")
library(Matrix)
library(ggplot2)
library(dplyr)
library(ggrepel)
```

```{r}
colors = c('#a6cee3',
          '#1f78b4',
          '#b2df8a',
          '#33a02c',
          '#fb9a99',
          '#e31a1c',
          '#fdbf6f',
          '#ff7f00',
          '#cab2d6',
          '#6a3d9a',
          '#ffff99',
          '#b15928',
          "#8B8A5F",
          "#D26C6D",
          "#F2F4ED"
          )
```

```{r}
init_tol=0.01
res = readRDS(file=paste('/project2/mstephens/dongyue/poisson_mf/droplet/droplet_iteration_results/ebpmf_droplet_full_nonnegLF_inittol',init_tol,'_iter60.rds',sep=''))
dim(res$fit_flash$L.pm)
p = structure_plot_general(res$fit_flash$L.pm,res$fit_flash$F.pm,samples$tissue,
                       LD=T,remove_l0f0 = T,title = 'EBPMF',
                       n_samples = 2000,K=7,print_plot = T)
ggsave(filename = "output/plots/droplet_k7_structure_plot.pdf", plot = p, device = "pdf", width = 10, height = 4.5, units = "in")
# structure_plot_general(res$fit_flash$L.pm,res$fit_flash$F.pm,samples$tissue,
#                        LD=T,remove_l0f0 = T,title = 'EBPMF',
#                        n_samples = 2000,K=9,print_plot = F)
# structure_plot_general(res$fit_flash$L.pm,res$fit_flash$F.pm,samples$tissue,
#                        LD=T,remove_l0f0 = T,title = 'EBPMF',
#                        n_samples = 2000,K=12,print_plot = F,colors = colors)

```

Zoom in rare cell factors

```{r}
kset=11

ldf = my_ldf(res$fit_flash$L.pm[,-c(1,2)],res$fit_flash$F.pm[,-c(1,2)])
Lhat = ldf$l%*%diag(ldf$d)
rare_idx = which(samples$tissue %in% c("Goblet","Ionocyte", "Neuroendocrine","Tuft"))
Lhat = Lhat[rare_idx,1:kset]
Fhat = matrix(1,nrow=3,ncol=ncol(Lhat))
if(is.null(colnames(Lhat))){
  colnames(Lhat) <- paste0("k",1:ncol(Lhat))
}
fit_list     <- list(L = Lhat,F = Fhat)
class(fit_list) <- c("multinom_topic_model_fit", "list")
rare_cell_type = as.factor(as.character(samples$tissue[rare_idx]))
p = structure_plot(fit_list,grouping = rare_cell_type,
                    loadings_order = 'embed',
                    n = 2000,gap = 5,colors=colors,verbose=F) +
  labs(y = "loading",color = "dim",fill = "dim") +ggtitle("EBPMF, rare cell types")
ggsave(filename = "output/plots/droplet_k11_structure_plot_rare.pdf", plot = p, device = "pdf", width = 10, height = 4.5, units = "in")
```

## Abandunt vs rare?

```{r}
abt_rare_ct = as.character(samples$tissue)
abt_rare_ct[abt_rare_ct%in%c('Club','Basal','Ciliated')] = 'abundant'
abt_rare_ct[abt_rare_ct!='abundant'] = 'rare'
abt_rare_ct = as.factor(abt_rare_ct)

structure_plot_general(res$fit_flash$L.pm,res$fit_flash$F.pm,abt_rare_ct,
                       LD=T,remove_l0f0 = T,title = 'EBPMF',
                       n_samples = 2000,K=7,print_plot = F)
```



## show some factor plots

```{r}
ldf = my_ldf(res$fit_flash$L.pm[,-c(1,2)],res$fit_flash$F.pm[,-c(1,2)])
```

The topic 1 is for ciliated cells. 

```{r}
k = 1
mean_exp_log = log10(colMeans(counts/rowSums(counts)))
colnames(counts)[order(ldf$f[,k],decreasing = T)[1:100]]

# Create a data frame for plotting
plot_data <- data.frame(
    mean_exp_log = mean_exp_log,
    gene_exp_program = ldf$f[, k],
    gene = colnames(counts)
)

# Define custom colors
custom_colors <- c("Cdhr3" = "black", "Foxj1" = "black", "Ccdc153" = "black", "Ccdc113" = "black", "Mlf1" = "black")


# Add a new column to the dataset with the appropriate colors
plot_data <- plot_data %>%
  mutate(color = ifelse(gene %in% names(custom_colors), custom_colors[gene], "grey80")) %>%
  arrange(desc(color))  # Sort the data frame based on the 'color' column

# Create the ggplot2 plot
p <- ggplot(plot_data, aes(x = mean_exp_log, y = gene_exp_program, color = color, shape = color)) +
    geom_point(size = 1) +
    scale_color_identity() +
    scale_shape_manual(values = c("grey80" = 19, "black" = 19, "red" = 19)) +
    theme_minimal() +
    labs(x = "Gene mean expression(log10)", y = "Gene expression program") +
    theme(legend.position = "none") +
    geom_text_repel(
        aes(label = gene),
        data = filter(plot_data, gene %in% c("Cdhr3", "Foxj1", "Ccdc153", "Ccdc113", "Mlf1")),
        size = 3,
        color='black'
    )+ggtitle('Factor 1')

# Print the plot
print(p)
ggsave(filename = paste('output/plots/droplet_k',k,'_factor.pdf',sep=''), plot = p, device = "pdf", width = 8, height = 5, units = "in")
# plot(mean_exp_log,ldf$f[,k],col='grey80',pch=19,xlab='Gene mean expression(log10)',ylab='Gene expression program')
# # Foxj1, Ccdc153, Ccdc113, Mlf1, Lztfl1
# 
# lines(mean_exp_log[which(colnames(counts)=='Cdhr3')],
#       ldf$f[which(colnames(counts)=='Cdhr3'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Foxj1')],
#       ldf$f[which(colnames(counts)=='Foxj1'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Ccdc153')],
#       ldf$f[which(colnames(counts)=='Ccdc153'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Ccdc113')],
#       ldf$f[which(colnames(counts)=='Ccdc113'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Mlf1')],
#       ldf$f[which(colnames(counts)=='Mlf1'),k],type='p',col=4,pch=19)
# # lines(mean_exp_log[which(colnames(counts)=='Lztfl1')],
# #       ldf$f[which(colnames(counts)=='Lztfl1'),k],type='p',col=4,pch=19)
# 
# text(mean_exp_log[which(colnames(counts)=='Cdhr3')]+0.2,ldf$f[which(colnames(counts)=='Cdhr3'),k],label='Cdhr3',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Foxj1')]+0.2,ldf$f[which(colnames(counts)=='Foxj1'),k],label='Foxj1',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Ccdc153')]+0.3,ldf$f[which(colnames(counts)=='Ccdc153'),k],label='Ccdc153',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Ccdc113')]-0.3,ldf$f[which(colnames(counts)=='Ccdc113'),k],label='Ccdc113',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Mlf1')]+0.2,ldf$f[which(colnames(counts)=='Mlf1'),k],label='Mlf1',cex=0.8)
```

topic 2 is for Basal cell

```{r}
k = 2
colnames(counts)[order(my_ldf(res$fit_flash$L.pm[,-c(1,2)],res$fit_flash$F.pm[,-c(1,2)])$f[,k],decreasing = T)[1:10]]


plot_data <- data.frame(
    mean_exp_log = mean_exp_log,
    gene_exp_program = ldf$f[, k],
    gene = colnames(counts)
)

# Define custom colors
custom_colors <- c("Krt5" = "black", "Krt17" = "black", "Aqp3" = "black", "Dapl1" = "black", "Hspa1a" = "black","Trp63" = "black")


# Add a new column to the dataset with the appropriate colors
plot_data <- plot_data %>%
  mutate(color = ifelse(gene %in% names(custom_colors), custom_colors[gene], "grey80")) %>%
  arrange(desc(color))  # Sort the data frame based on the 'color' column

# Create the ggplot2 plot
p <- ggplot(plot_data, aes(x = mean_exp_log, y = gene_exp_program, color = color, shape = color)) +
    geom_point(size = 1) +
    scale_color_identity() +
    scale_shape_manual(values = c("grey80" = 19, "black" = 19, "red" = 19)) +
    theme_minimal() +
    labs(x = "Gene mean expression(log10)", y = "Gene expression program") +
    theme(legend.position = "none") +
    geom_text_repel(
        aes(label = gene),
        data = filter(plot_data, gene %in% c("Krt5", "Krt17", "Aqp3", "Dapl1", "Hspa1a",'Trp63')),
        size = 3,
        color = "black"
    )+ggtitle(paste('Factor ',k,sep=''))

# Print the plot
print(p)
ggsave(filename = paste('output/plots/droplet_k',k,'_factor.pdf',sep=''), plot = p, device = "pdf", width = 8, height = 5, units = "in")
# plot(mean_exp_log,ldf$f[,k],col='grey80',pch=19,xlab='Gene mean expression(log10)',ylab='Gene expression program')
# #identified_points <- identify(mean_exp_log,ldf$f[,k+2])
# #colnames(counts)[identified_points]
# 
# #aqp3, kRT5, dAPL1, hSPA1A,tRP63
# 
# lines(mean_exp_log[which(colnames(counts)=='Krt5')],
#       ldf$f[which(colnames(counts)=='Krt5'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Krt17')],
#       ldf$f[which(colnames(counts)=='Krt17'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Aqp3')],
#       ldf$f[which(colnames(counts)=='Aqp3'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Dapl1')],
#       ldf$f[which(colnames(counts)=='Dapl1'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Hspa1a')],
#       ldf$f[which(colnames(counts)=='Hspa1a'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Trp63')],
#       ldf$f[which(colnames(counts)=='Trp63'),k],type='p',col=4,pch=19)
# 
# text(mean_exp_log[which(colnames(counts)=='Krt5')]+0.2,ldf$f[which(colnames(counts)=='Krt5'),k],label='Krt5',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Krt17')]+0.2,ldf$f[which(colnames(counts)=='Krt17'),k],label='Krt17',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Aqp3')]+0.2,ldf$f[which(colnames(counts)=='Aqp3'),k],label='Aqp3',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Dapl1')]+0.2,ldf$f[which(colnames(counts)=='Dapl1'),k],label='Dapl1',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Hspa1a')]+0.3,ldf$f[which(colnames(counts)=='Hspa1a'),k],label='Hspa1a',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Trp63')]+0.2,ldf$f[which(colnames(counts)=='Trp63'),k],label='Trp63',cex=0.8)
```

topic 3 is for tuft, neuroendocrine, and ionocyte.?? mainly for tuft

```{r}
# Rgs13 (tuft cells), Dclk 1, Chga, Ascl3
k = 3
colnames(counts)[order(ldf$f[,k],decreasing = T)[1:50]]
plot_data <- data.frame(
    mean_exp_log = mean_exp_log,
    gene_exp_program = ldf$f[, k],
    gene = colnames(counts)
)

# Define custom colors
custom_colors <- c("Dclk1" = "black", "Rgs13" = "black")


# Add a new column to the dataset with the appropriate colors
plot_data <- plot_data %>%
  mutate(color = ifelse(gene %in% names(custom_colors), custom_colors[gene], "grey80")) %>%
  arrange(desc(color))  # Sort the data frame based on the 'color' column

# Create the ggplot2 plot
p <- ggplot(plot_data, aes(x = mean_exp_log, y = gene_exp_program, color = color, shape = color)) +
    geom_point(size = 1) +
    scale_color_identity() +
    scale_shape_manual(values = c("grey80" = 19, "black" = 19, "red" = 19)) +
    theme_minimal() +
    labs(x = "Gene mean expression(log10)", y = "Gene expression program") +
    theme(legend.position = "none") +
    geom_text_repel(
        aes(label = gene),
        data = filter(plot_data, gene %in% c("Dclk1", "Rgs13")),
        size = 3,
        color = "black"
    )+ggtitle(paste('Factor ',k,sep=''))

# Print the plot
print(p)
ggsave(filename = paste('output/plots/droplet_k',k,'_factor.pdf',sep=''), plot = p, device = "pdf", width = 8, height = 5, units = "in")

# plot(mean_exp_log,ldf$f[,k],col='grey80',pch=19,xlab='Gene mean expression(log10)',ylab='Gene expression program')
# #identified_points <- identify(mean_exp_log,ldf$f[,k+2])
# #colnames(counts)[identified_points]
# 
# lines(mean_exp_log[which(colnames(counts)=='Dclk1')],
#       ldf$f[which(colnames(counts)=='Dclk1'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Rgs13')],
#       ldf$f[which(colnames(counts)=='Rgs13'),k],type='p',col=4,pch=19)
# # lines(mean_exp_log[which(colnames(counts)=='Ascl3')],
# #       ldf$f[which(colnames(counts)=='Ascl3'),k],type='p',col=4,pch=19)
```

topic 11 is for globlet

```{r}
# Gp2: marker
# Goblet-1 cells are enriched for the expression of genes encoding key mucosal proteins (Tff1, Tff2 and Muc5b19, FDR <0.001, LRT) and secretory regulators (for example, Lman1l or P2rx427, FDR <0.1, LRT).
# Goblet-2 cells specifically express Dcpp1, Dcpp2 and Dcpp3, orthologues of ZG16B, which codes for a lectin-like secreted protein that aggregates bacteria28, and Lipf, a secreted gastric lipase that hydrolyses triglycerides.
k = 11
colnames(counts)[order(ldf$f[,k],decreasing = T)[1:50]]

colnames(counts)[order(ldf$f[,k],decreasing = T)[1:50]]
plot_data <- data.frame(
    mean_exp_log = mean_exp_log,
    gene_exp_program = ldf$f[, k],
    gene = colnames(counts)
)

# Define custom colors
custom_colors <- c("Gp2" = "black", "Tff1" = "red","Tff2" = "red","Lman1l" = "red","Muc5b" = "red",
                   "Dcpp1" = "blue","Dcpp2" = "blue","Dcpp3" = "blue","Lipf" = "blue")


# Add a new column to the dataset with the appropriate colors
plot_data <- plot_data %>%
  mutate(color = ifelse(gene %in% names(custom_colors), custom_colors[gene], "grey80")) %>%
  arrange(desc(color))  # Sort the data frame based on the 'color' column

# Create the ggplot2 plot
p <- ggplot(plot_data, aes(x = mean_exp_log, y = gene_exp_program, color = color, shape = color)) +
    geom_point(size = 1) +
    scale_color_identity() +
    scale_shape_manual(values = c("grey80" = 19, "black" = 19, "red" = 19,"blue" = 19)) +
    theme_minimal() +
    labs(x = "Gene mean expression(log10)", y = "Gene expression program") +
    theme(legend.position = "none") +
    geom_text_repel(
        aes(label = gene),
        data = filter(plot_data, gene %in% c("Gp2","Tff1","Tff2","Lman1l","Muc5b","Dcpp1","Dcpp2","Dcpp3","Lipf")),
        size = 3,
        color = "black"
    )+ggtitle(paste('Factor ',k,sep=''))

# Print the plot
print(p)
ggsave(filename = paste('output/plots/droplet_k',k,'_factor.pdf',sep=''), plot = p, device = "pdf", width = 8, height = 5, units = "in")


# plot(mean_exp_log,ldf$f[,k],col='grey80',pch=19,xlab='Gene mean expression(log10)',ylab='Gene expression program')
# #identified_points <- identify(mean_exp_log,ldf$f[,k+2])
# #colnames(counts)[identified_points]
# 
# lines(mean_exp_log[which(colnames(counts)=='Gp2')],
#       ldf$f[which(colnames(counts)=='Gp2'),k],type='p',col=4,pch=19)
# text(mean_exp_log[which(colnames(counts)=='Gp2')]+0.2,ldf$f[which(colnames(counts)=='Gp2'),k],label='Gp2',cex=0.8)
# 
# lines(mean_exp_log[which(colnames(counts)=='Tff1')],ldf$f[which(colnames(counts)=='Tff1'),k],type='p',col=4,pch=19)
# text(mean_exp_log[which(colnames(counts)=='Tff1')]+0.2,ldf$f[which(colnames(counts)=='Tff1'),k],label='Tff1',cex=0.8)
# lines(mean_exp_log[which(colnames(counts)=='Tff2')],ldf$f[which(colnames(counts)=='Tff2'),k],type='p',col=4,pch=19)
# text(mean_exp_log[which(colnames(counts)=='Tff2')]+0.2,ldf$f[which(colnames(counts)=='Tff2'),k],label='Tff2',cex=0.8)
# lines(mean_exp_log[which(colnames(counts)=='Lman1l')], ldf$f[which(colnames(counts)=='Lman1l'),k],type='p',col=4,pch=19)
# text(mean_exp_log[which(colnames(counts)=='Lman1l')]+0.2,ldf$f[which(colnames(counts)=='Lman1l'),k],label='Lman1l',cex=0.8)
# lines(mean_exp_log[which(colnames(counts)=='Muc5b')], ldf$f[which(colnames(counts)=='Muc5b'),k],type='p',col=4,pch=19)
# text(mean_exp_log[which(colnames(counts)=='Muc5b')]+0.2,ldf$f[which(colnames(counts)=='Muc5b'),k],label='Muc5b',cex=0.8)
# 
# lines(mean_exp_log[which(colnames(counts)=='Dcpp1')],
#       ldf$f[which(colnames(counts)=='Dcpp1'),k],type='p',col=2,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Dcpp2')],
#       ldf$f[which(colnames(counts)=='Dcpp2'),k],type='p',col=2,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Dcpp3')],
#       ldf$f[which(colnames(counts)=='Dcpp3'),k],type='p',col=2,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Lipf')],
#       ldf$f[which(colnames(counts)=='Lipf'),k],type='p',col=2,pch=19)
# text(mean_exp_log[which(colnames(counts)=='Dcpp1')]+0.2,ldf$f[which(colnames(counts)=='Dcpp1'),k],label='Dcpp1',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Dcpp2')]+0.2,ldf$f[which(colnames(counts)=='Dcpp2'),k],label='Dcpp2',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Dcpp3')]+0.2,ldf$f[which(colnames(counts)=='Dcpp3'),k],label='Dcpp3',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Lipf')]+0.2,ldf$f[which(colnames(counts)=='Lipf'),k],label='Lipf',cex=0.8)
```

topic 6 specific to neuroendocrine

```{r}
k = 6
colnames(counts)[order(ldf$f[,k],decreasing = T)[1:50]]

plot_data <- data.frame(
    mean_exp_log = mean_exp_log,
    gene_exp_program = ldf$f[, k],
    gene = colnames(counts)
)

# Define custom colors
custom_colors <- c("Chga" = "black", "Calca" = "red","Cxcl13" = "red",
                   "Ly6h"="red","Pkib"="red","Tmem163"="red",  
                   "Ddc"="red","Pcsk1" ="red",
                   "Tcerg1l"="red",  "Scg2"="red",     "Scn9a"="red",   "Stmn3"="red",    "Chgb"="red")


# Add a new column to the dataset with the appropriate colors
plot_data <- plot_data %>%
  mutate(color = ifelse(gene %in% names(custom_colors), custom_colors[gene], "grey80")) %>%
  arrange(desc(color))  # Sort the data frame based on the 'color' column

# Create the ggplot2 plot
p <- ggplot(plot_data, aes(x = mean_exp_log, y = gene_exp_program, color = color, shape = color)) +
    geom_point(size = 1) +
    scale_color_identity() +
    scale_shape_manual(values = c("grey80" = 19, "black" = 19, "red" = 19)) +
    theme_minimal() +
    labs(x = "Gene mean expression(log10)", y = "Gene expression program") +
    theme(legend.position = "none") +
    geom_text_repel(
        aes(label = gene),
        data = filter(plot_data, gene %in% c("Chga", "Calca","Cxcl13","Ly6h" , "Pkib"  ,   "Tmem163" , "Ddc"    ,  "Pcsk1","Tcerg1l" , "Scg2" ,  "Scn9a" ,   "Stmn3"  ,  "Chgb" )),
        size = 3,
        color = "black",
        max.overlaps = 100
    )+ggtitle(paste('Factor ',k,sep=''))

# Print the plot
print(p)
ggsave(filename = paste('output/plots/droplet_k',k,'_factor.pdf',sep=''), plot = p, device = "pdf", width = 8, height = 5, units = "in")


# plot(mean_exp_log,ldf$f[,k],col='grey80',pch=19,xlab='Gene mean expression(log10)',ylab='Gene expression program')
# #identified_points <- identify(mean_exp_log,ldf$f[,k+2])
# #colnames(counts)[identified_points]
# 
# lines(mean_exp_log[which(colnames(counts)=='Chga')],ldf$f[which(colnames(counts)=='Chga'),k],type='p',col=4,pch=19)
# text(mean_exp_log[which(colnames(counts)=='Chga')]+0.2,ldf$f[which(colnames(counts)=='Chga'),k],label='Chga',cex=0.8)

```

topic 9 specific to ionocyte

```{r}
k = 9
colnames(counts)[order(ldf$f[,k],decreasing = T)[1:50]]


# Create a data frame for plotting
plot_data <- data.frame(
    mean_exp_log = mean_exp_log,
    gene_exp_program = ldf$f[, k],
    gene = colnames(counts)
)

# Define custom colors
custom_colors <- c("Foxi1" = "black", "Ascl3" = "black", "Cftr" = "black",
                   "Gm933"="red","Moxd1"="red",   "Stap1"="red",    "Syn2"="red",     "Atp6v0d2"="red", "Coch" ="red", "Tfcp2l1"="red",
                   "Nrip3"="red")

plot_data <- plot_data %>%
  mutate(color = ifelse(gene %in% names(custom_colors), custom_colors[gene], "grey80")) %>%
  arrange(desc(color))  # Sort the data frame based on the 'color' column

# Create the ggplot2 plot
p <- ggplot(plot_data, aes(x = mean_exp_log, y = gene_exp_program, color = color, shape = color)) +
    geom_point(size = 1) +
    scale_color_identity() +
    scale_shape_manual(values = c("grey80" = 19, "black" = 19, "red" = 19)) +
    theme_minimal() +
    labs(x = "Gene mean expression(log10)", y = "Gene expression program") +
    theme(legend.position = "none") +
    geom_text_repel(
        aes(label = gene),
        data = filter(plot_data, gene %in% c("Foxi1", "Ascl3", "Cftr","Gm933","Moxd1",   "Stap1",    "Syn2",     "Atp6v0d2", "Coch" , "Tfcp2l1","Nrip3")),
        size = 3,
        color='black',
        max.overlaps = 100
    )+ggtitle(paste('Factor ',k,sep=''))

# Print the plot
print(p)
ggsave(filename = paste('output/plots/droplet_k',k,'_factor.pdf',sep=''), plot = p, device = "pdf", width = 8, height = 5, units = "in")

# plot(mean_exp_log,ldf$f[,k],col='grey80',pch=19,xlab='Gene mean expression(log10)',ylab='Gene expression program')
# 
# lines(mean_exp_log[which(colnames(counts)=='Foxi1')],ldf$f[which(colnames(counts)=='Foxi1'),k],type='p',col=4,pch=19)
# lines(mean_exp_log[which(colnames(counts)=='Ascl3')],ldf$f[which(colnames(counts)=='Ascl3'),k],type='p',col=4,pch=19)
# text(mean_exp_log[which(colnames(counts)=='Foxi1')]+0.2,ldf$f[which(colnames(counts)=='Foxi1'),k],label='Foxi1',cex=0.8)
# text(mean_exp_log[which(colnames(counts)=='Ascl3')]+0.2,ldf$f[which(colnames(counts)=='Ascl3'),k],label='Ascl3',cex=0.8)
# 
# lines(mean_exp_log[which(colnames(counts)=='Cftr')],ldf$f[which(colnames(counts)=='Cftr'),k],type='p',col=2,pch=19)
# text(mean_exp_log[which(colnames(counts)=='Cftr')]+0.2,ldf$f[which(colnames(counts)=='Cftr'),k],label='Cftr',cex=0.8)

```


k = 7 is for club cells? 

```{r}
k = 7
colnames(counts)[order(ldf$f[,k],decreasing = T)[1:50]]
plot(mean_exp_log,ldf$f[,k],col='grey80',pch=19,xlab='Gene mean expression(log10)',ylab='Gene expression program')
#identified_points <- identify(mean_exp_log,ldf$f[,k+2])
#colnames(counts)[identified_points]

lines(mean_exp_log[which(colnames(counts)=='Nfia')],
      ldf$f[which(colnames(counts)=='Nfia'),k],type='p',col=4,pch=19)
text(mean_exp_log[which(colnames(counts)=='Nfia')]+0.2,ldf$f[which(colnames(counts)=='Nfia'),k],label='Nfia',cex=0.8)
```

## pulse seq 66k cells, gene selected as union of scry:deviance and seurat

```{r}
init_tol=0.01
res = readRDS(file=paste('/project2/mstephens/dongyue/poisson_mf/droplet_66k/iteration_results/ebpmf_pulse_seq_5kgene_nonnegLF_inittol',init_tol,'_iter50.rds',sep=''))
```

```{r}
dat =readRDS('/project2/mstephens/dongyue/poisson_mf/droplet_66k/pulse_seq_5kgene.rds')
cell.type <- sapply(strsplit(rownames(dat), "_"), `[[`, 5)
time.point <- sapply(strsplit(rownames(dat), "_"), `[[`, 2)

cell.type <- ifelse(
  cell.type == "Basal", paste0("Basal (", time.point, ")"), cell.type
)

cell.type <- factor(cell.type, levels = c(
  "Basal (Tp0)",
  "Basal (Tp30)",
  "Basal (Tp60)",
  "Club",
  "Club (hillock-associated)",
  "Ciliated",
  "Goblet.progenitor",
  "Goblet.1",
  "Goblet.2",
  "Tuft.progenitor",
  "Tuft.1",
  "Tuft.2",
  "Neuroendocrine",
  "Ionocyte",
  "Proliferating"
))
```

```{r}
colors = randomcoloR::distinctColorPalette(k=16,runTsne = T)
structure_plot_general(res$fit_flash$L.pm,res$fit_flash$F.pm,cell.type,
                       LD=T,remove_l0f0 = T,title = 'EBPMF',
                       n_samples = 2000,K=16,print_plot = F,colors = colors)

structure_plot_general(res$fit_flash$L.pm,res$fit_flash$F.pm,cell.type,
                       LD=F,remove_l0f0 = T,title = 'EBPMF',
                       n_samples = 2000,K=16,print_plot = F,colors = colors)

```


Only show those for rare cell types

```{r}
rare_idx = which(cell.type%in%c("Goblet.progenitor",
  "Goblet.1",
  "Goblet.2",
  "Tuft.progenitor",
  "Tuft.1",
  "Tuft.2",
  "Neuroendocrine",
  "Ionocyte",
  "Proliferating"))

ldf = my_ldf(res$fit_flash$L.pm[,-c(1,2)],res$fit_flash$F.pm[,-c(1,2)])
Lhat = ldf$l%*%diag(ldf$d)
Lhat = Lhat[rare_idx,]
Fhat = matrix(1,nrow=3,ncol=ncol(Lhat))
if(is.null(colnames(Lhat))){
  colnames(Lhat) <- paste0("k",1:ncol(Lhat))
}
fit_list     <- list(L = Lhat,F = Fhat)
class(fit_list) <- c("multinom_topic_model_fit", "list")
structure_plot(fit_list,grouping = as.factor(as.character(cell.type[rare_idx])),
                      n = 2000,gap = 40,colors=colors,verbose=F) +
    labs(y = "loading",color = "dim",fill = "dim") 
```


Heatmap of the loadings

```{r}
FF = ldf$l
cell.type <- sapply(strsplit(rownames(dat), "_"), `[[`, 5)
cell.type <- factor(cell.type, levels = sort(unique(cell.type))[c(
  1, 4, 3, 2, 7, 5, 6, 13, 11, 12, 9, 8, 10
)])

# Downsample the number of cells and sort them using tSNE.
set.seed(666)
cell.idx <- numeric(0)
cell.types <- levels(cell.type)
for (i in 1:length(cell.types)) {
  which.idx <- which(cell.type == cell.types[i])
  # Downsample common cell types.
  if (length(which.idx) > 100) {
    which.idx <- sample(which.idx, 100)
  }
  # Don't include rare cell types.
  if (length(which.idx) > 20) {
    # Sort using tsne.
    tsne.res <- Rtsne::Rtsne(
      FF[which.idx, ],
      dims = 1,
      pca = FALSE,
      normalize = FALSE,
      perplexity = min(100, floor((length(which.idx) - 1) / 3) - 1),
      theta = 0.1,
      max_iter = 1000,
      eta = 200,
      check_duplicates = FALSE
    )$Y[, 1]
    which.idx <- which.idx[order(tsne.res)]
    cell.idx <- c(cell.idx, which.idx)
  }
}

cell.type <- cell.type[cell.idx]
cell.type <- droplevels(cell.type)

ebmf.FF <- FF[cell.idx, ]
make.heatmap.tib <- function(FF) {
  tib <- as_tibble(scale(FF, center = FALSE, scale = apply(FF, 2, max))) %>%
    mutate(Cell.type = cell.type) %>%
    arrange(Cell.type) %>%
    mutate(Cell.idx = row_number())
  
  tib <- tib %>%
    pivot_longer(
      -c(Cell.idx, Cell.type),
      names_to = "Factor",
      values_to = "Loading",
      values_drop_na = TRUE
    ) %>%
    mutate(Factor = as.numeric(str_extract(Factor, "[0-9]+")))
  
  return(tib)
}

ebmf.tib <- make.heatmap.tib(ebmf.FF)
heatmap.tib <- ebmf.tib %>% mutate(Method = "EBPMF")
tib <- heatmap.tib %>%
  group_by(Cell.type, Cell.idx) %>%
  summarize()

cell_type_breaks <- c(1, which(tib$Cell.type[-1] != tib$Cell.type[-nrow(tib)]))
label_pos <- cell_type_breaks / 2 + c(cell_type_breaks[-1], nrow(tib)) / 2

plt <- ggplot(heatmap.tib, aes(x = Factor, y = -Cell.idx, fill = Loading)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "firebrick") +
  labs(y = "") +
  scale_y_continuous(breaks = -label_pos,
                     minor_breaks = NULL,
                     labels = levels(cell.type)) +
  theme_minimal() +
  geom_hline(yintercept = -cell_type_breaks, size = 0.1) +
  facet_wrap(~Method, ncol = 1) +
  theme(legend.position = "none", 
        strip.text = element_text(size = 16))

plt
```


## pulse seq 66k cells, gene selected using rliger

```{r}
init_tol=0.1
res = readRDS(file=paste('/project2/mstephens/dongyue/poisson_mf/droplet_66k/iteration_results/ebpmf_pulse_seq_3kgene_nonnegLF_inittol',init_tol,'_iter50.rds',sep=''))
```

```{r}
dat =readRDS('/project2/mstephens/dongyue/poisson_mf/droplet_66k/pulse_seq_3kgene_rliger.rds')
cell.type <- sapply(strsplit(rownames(dat), "_"), `[[`, 5)
time.point <- sapply(strsplit(rownames(dat), "_"), `[[`, 2)

cell.type <- ifelse(
  cell.type == "Basal", paste0("Basal (", time.point, ")"), cell.type
)

cell.type <- factor(cell.type, levels = c(
  "Basal (Tp0)",
  "Basal (Tp30)",
  "Basal (Tp60)",
  "Club",
  "Club (hillock-associated)",
  "Ciliated",
  "Goblet.progenitor",
  "Goblet.1",
  "Goblet.2",
  "Tuft.progenitor",
  "Tuft.1",
  "Tuft.2",
  "Neuroendocrine",
  "Ionocyte",
  "Proliferating"
))
```

```{r}
dim(res$fit_flash$L.pm)
colors = randomcoloR::distinctColorPalette(k=16,runTsne = T)
structure_plot_general(res$fit_flash$L.pm,res$fit_flash$F.pm,cell.type,
                       LD=T,remove_l0f0 = T,title = 'EBPMF',
                       n_samples = 2000,K=13,print_plot = F,colors = colors)
```


Only show those for rare cell types

```{r}
rare_idx = which(cell.type%in%c("Goblet.progenitor",
  "Goblet.1",
  "Goblet.2",
  "Tuft.progenitor",
  "Tuft.1",
  "Tuft.2",
  "Neuroendocrine",
  "Ionocyte",
  "Proliferating"))

ldf = my_ldf(res$fit_flash$L.pm[,-c(1,2)],res$fit_flash$F.pm[,-c(1,2)])
Lhat = ldf$l%*%diag(ldf$d)
Lhat = Lhat[rare_idx,]
Fhat = matrix(1,nrow=3,ncol=ncol(Lhat))
if(is.null(colnames(Lhat))){
  colnames(Lhat) <- paste0("k",1:ncol(Lhat))
}
fit_list     <- list(L = Lhat,F = Fhat)
class(fit_list) <- c("multinom_topic_model_fit", "list")
structure_plot(fit_list,grouping = as.factor(as.character(cell.type[rare_idx])),
                      n = 2000,gap = 40,colors=colors,verbose=F) +
    labs(y = "loading",color = "dim",fill = "dim") 
```

