---
title: "hOMEPAGe"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: false
editor_options:
  chunk_output_type: console
---

## Poisson Matrix Factorization


### Analysis

* [Negative Binomial and Poisson log-normal mean-var relationship](NB_PLN_mean_var.html)
* [mean variance relationship in real single cell data](mean_var_relationship_in_real_sc_data.html)
* [attempt to reduce run time](attempt_to_speed_up_add_greedy.html)
* [memory usage analysis](splitting_PMF_memory_issue.html)
* [size of sigma2 on convergence speed](size_of_sigma2_on_convergence.html)
* [fit K=1 on pbmc data](run_splitting_PMF_K1.html)
* [point exponential and unimodal nonnegative]
* [ebpm normal prior init in PMF](vga_init_in_ebpmf.html)
* [ebpmf vs topic model](ebpmf_nmf_diff.html)
* [Poisson deviance](poisson_deviance.html)
* [Multiplicative vs additive](multiplicative_additive.html)
* [smallman example](smallman_example.html)


### Simulation 

* [ebpmf log with smoothed factor](ebpmf_smooth_factor_example.Rmd)
* [Splitting method in PMF, simple simulation example](PMF_splitting.html)
* [Simulation based on PBMC data, 3 cell types](simulation_based_on_pbmc_3cell.html)
* [Simulation based on PBMC data, 3 cell types, 1 init backfitting](splitting_PMF_1backfitinit.html)
* [Simulation pbmc data from fasttopics, nonnegtive L and F](simu_pbmc_fasttpoics_nonnegLF.html)
* [Simulation data from fitted model on pbmc data in fasttopics, nonnegtive L and F](simu_study_of_init_variance.html)
* [Simulation data from fitted model on pbmc data in fasttopics, nonnegtive L and F, added more models](simu_study_pbmc_truevarlarge.html)
* [Simulation data from fitted model on pbmc data in fasttopics, nonnegtive L and F, true sigma2 = 0](simu_study_pbmc_truevar0.html)

### On text data

#### SLA data set

* [EBPMF does not find factors?](fit_ebpmf_sla.html)
* [Pre-processing SLA data](sla_data.html)
* [EBPMF does not find factors?](fit_ebpmf_sla_2000.html)
* [log transformation + EBNMF on SLA data](sla_ebmf_intercept.html)
* [ebpmf init using liklihood expansion](ebpmf_init_likexp.html)
* [methods fit on sla 2000 data](fit_ebpmf_sla_2000.html)
* [methods fit on sla full data, filter out words in less than 5 doc](fit_ebpmf_sla_full_nonneg.html)
* [methods fit on sla full data, filter out words in less than 3 doc](fit_ebpmf_sla_full_nonneg2.html)
* [sla flash init](sla_flash_init.html)

### On pbmc3k

* [pbmc data in fasttopics package, flash, glmpca, and topic model fit](pbmc_infasttopics_othermethods.html)
* [pbmc3k data 10X, flash, glmpca, and topic model fit](pbmc3k_10X_othermethods.html)
* [pbmc3k 10X result](pbmc3k_10X_result2.html)
* [pbmc in fasttopic results](pbmc3k_infasttopic_result2.html)
* [pbmc in fasttopic,effect of initialization variances](pbmc_infasttopics_init_vga_effect.html)
* [pbmc in fasttopic, 3-5 cell types for simulation]()

Below are some analysis I have done, before putting $l_0,f_0$ as part of flash, and estimating $f_0$.

* [run PMF on PBMC data, 3 cell types](run_PMF_on_pbmc_3cells.html)
* [run PMF on PBMC data, 3 cell types, nonnegative loadings](run_PMF_on_pbmc_3cells_nonnegativeL_more_iterations.html)
* [run PMF on PBMC data, 3 cell types, nonnegative loadings and factors](run_PMF_on_pbmc_3cells_nonnegativeLF_more_iterations.html)
* [Compare VGA iterations on PBMC data, 3 cell types](compare_vga_iterations.html)
* [Compare VGA iterations on PBMC data, full dataset](compare_vga_iterations_pbmc_full.html)
* [run PMF on PBMC data](run_PMF_on_pbmc_more_iterations.html)
* [run PMF on PBMC data,nonnegative Loadings](run_PMF_on_pbmc_nonnegativeL_more_iterations.html)
* [run PMF on PBMC data,nonnegative Loadings and factors](run_PMF_on_pbmc_nonnegativeLF_more_iterations.html)
* [library size distribution](library_size_distribution.html)
* [gene specific variance and gene expression](gene_variance_expression.html)
* [Compare latent M from splitting and glmpca](compare_latent_M_splitting_glmpca.html)
* [Run flash on latent M from glmpca](run_flash_on_latent_M_glmpca.html)
* [How do results differ over iterations? When can we stop iterations?](pbmc3k_per_10_iter_plot.html)
* [Use GLMPCA as init](pbmc3k_glmpca_init_for_ebpmf.html)
* [estimate f0, nonnegative LF case](pbmc3k_nonnegLF_updatef0.html)


### On Droplet

* [droplet Montoro results](run_PMF_on_droplet.html)
* [droplet Montoro results, no filtering of genes](droplet_results_full.html)

### On PBMC purified

* [PBMC purified result](pbmc_purified_results.html)

### On PBMC 68K

* [PBMC 68K result](pbmc68k_reults.html)

### On Pancreas data

## Poisson mean 

* [Poisson log normal](poisson_log_normal.html)
* [NB JJ lower bound](neg_binom_mean_lower_bound.html)
* [NB PG augmentation](neg_binom_mean_polya_gamma.html)
* [NB approxiamation to Poisson](nb_approx_poisson.html)
* [Poisson mean $\log(1+\exp(\cdot))$ link](pois_mean_log1exp.html)
* [VGA objective function v(m) relationship](vga_objective_function.html)
* [Poisson mean splitting local optimum](poisson_mean_split_local_optimum.html)

### Penalized

* [Poisson mean penalized form, compound](pois_mean_penalized.html)
* [Poisson mean penalized form, inversion](pois_mean_penalized_inversion.html)

### Benchmarking

* [log link](log_link_benchmark_revised.html)
* [exp prior](exp_prior_benchmark.html)
* [gamma prior](gamma_prior_benchmark.html)
* [t prior](t_prior_benchmark_revise.html)

With Run time Comparison:

* [exp prior with runtime](exp_prior_runtime.html)
* [log prior with runtime](log_prior_runtime.html)

### Overdispersed model

* [generate data from splitting model](overdispersed_splitting.html)
* [generate data from negative binomial model](overdispersed_splitting_nb.html)
* [thoughts on the results](overdispersed_splitting_analysis.html)

## Poisson smoothing 

* [The splitting method smoothing overdispersed count data](poisson_nugget_two_step.html)
* [Poisson smooth splitting local optimum](poisson_smooth_split_local_optimum.html)

### Real data

* [Smooth RNA seq](splitting_smooth_RNAseq.html)

### Bechmarking smoothing methods



[Some old results](smoothing_simu_2022.html)

### Details

* [low count smooth ndwt](splitting_method_low_count_smoothing.html)


## Binomial 

* [Apply Splitting to Binomial Mean problem](binomial_mean_spliiting.html)
* [It does not work for binary data](bernoulli_GG_fail.html)
* [Apply Splitting to Binomial smooth problem](binomial_smooth_splitting.html)
* [is vga binom problem convex?](is_vga_binom_convex.Rmd)

## Normal mean penalty form

* [bound on (log g)''](bound_log_marginal.html)
* [log density of normal mixture](log_gmm_concave.html)
  + [log density of normal mixture, variance ratio](log_gmm_concave_ratio.html)
* [log density of laplace mixture](log_lmm_concave.html)
* [normal mean penalty plots](normal_mean_penalty.html)
* [Solve normal mean problem using optimization - compund and inversion penalty](normal_mean_penalized_optimization.html)
* [Solve normal mean problem using optimization - also estimate prior mean](normal_mean_penalized_optimization_prior_mean.html)
* [normal mean penalty plots, when $s^2$ depends on posterior mean](normal_mean_penalty_glm_simplified.html)
* [normal mean penalty plots, when $s^2$ depends on posterior mean, penalty exlcuding $\log(2\pi s^2)/2$ ](remove_log_in_penalty.html)
* [where is the log term from in the penalty](log_term_in_penalty.html)

