---
title: "pbmc3k data in fasttopic package, result"
author: "DongyueXie"
date: "2023-02-16"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Introduction

This pbmc3k data is from fastTopics package, and is described as "These data are a selection of the reference transcriptome profiles generated via single-cell RNA sequencing (RNA-seq) of 10 bead-enriched subpopulations of PBMCs (Donor A), described in Zheng et al (2017)."


## Filter out genes

I filtered out genes that expressed in less than 10 cells, and no further preprocessing. The final data matrix is 

```{r}
library(fastTopics)
library(stm)
library(Matrix)
data(pbmc_facs)

## original data fit
counts = pbmc_facs$counts
counts_filtered = counts[,colSums(counts!=0)>10]
dim(counts_filtered)
```

```{r,eval=FALSE}
pbmc3k_sparse = ebpmf_log(counts_filtered,flash_control = list())
saveRDS(pbmc3k_sparse,'output/pbmc_fasttopics/pbmc3k_sparse.rds')
pbmc3k_nonnegL = ebpmf_log(counts_filtered,flash_control = list(ebnm.fn=c(ebnm::ebnm_point_exponential,ebnm::ebnm_point_normal),loadings_sign = 1))
saveRDS(pbmc3k_nonnegL,'output/pbmc_fasttopics/pbmc3k_nonnegL_pe.rds')
pbmc3k_nonnegLF = ebpmf_log(counts_filtered,flash_control = list(ebnm.fn=c(ebnm::ebnm_point_exponential,ebnm::ebnm_point_exponential),factors_sign=1,loadings_sign = 1))
saveRDS(pbmc3k_nonnegLF,'output/pbmc_fasttopics/pbmc3k_nonnegLF_pe.rds')

pbmc3k_nonnegL = ebpmf_log(counts_filtered,flash_control = list(ebnm.fn=c(ebnm::ebnm_unimodal_nonnegative,ebnm::ebnm_point_normal),loadings_sign = 1))
saveRDS(pbmc3k_nonnegL,'output/pbmc_fasttopics/pbmc3k_nonnegL_un.rds')
pbmc3k_nonnegLF = ebpmf_log(counts_filtered,flash_control = list(ebnm.fn=c(ebnm::ebnm_unimodal_nonnegative,ebnm::ebnm_unimodal_nonnegative),factors_sign=1,loadings_sign = 1))
saveRDS(pbmc3k_nonnegLF,'output/pbmc_fasttopics/pbmc3k_nonnegLF_un.rds')
```

## Model fitting

The model is the Empirical Bayes Poisson matrix factorization model, and $l_0$ is fixed at `log(rowMeans(Y))`, $f_0$ is initialized at `log(colSums(Y)/sum(l_0))`, then let the model estimates $f_0$ during iterations.


```{r}
summary_plot = function(res){
  plot(res$K_trace,ylab='',xlab='iterations',main='K over iterations',pch=20)
  plot(res$sigma2,ylab='',xlab='gene',main = 'gene specific variance',pch=20,col='grey70')
  plot(colSums(counts_filtered/rowSums(counts_filtered)),res$sigma2,xlab='gene expression',ylab='gene specific variance',pch=20,col='grey60')
  plot(colSums(counts_filtered==0)/nrow(counts_filtered),res$sigma2,xlab='gene expression sparsity',ylab='gene specific variance',pch = 20, col='grey50')
}
```

### Sparse loadings and factors

```{r}
source('code/poisson_STM/plot_factors.R')
pbmc3k_sparse = readRDS('output/pbmc_fasttopics/pbmc3k_sparse.rds')
```

```{r}
summary_plot(pbmc3k_sparse)
```

```{r}
plot.factors(pbmc3k_sparse$fit_flash,pbmc_facs$samples$subpop)
```

```{r}
f0_init = log(colSums(counts_filtered)/sum(rowMeans(counts_filtered)))
plot(pbmc3k_sparse$fit_flash$F.pm[,2],f0_init,xlab='f0 estimates',ylab='f0 init values',pch=20,col='grey70')
abline(a=0,b=1)
```


## Nonnegative loadings with point exponential prior, and sparse factors

```{r}
source('code/poisson_STM/structure_plot.R')
pbmc3k_nonnegL_pe = readRDS('output/pbmc_fasttopics/pbmc3k_nonnegL_pe.rds')
```

```{r}
summary_plot(pbmc3k_nonnegL_pe)
```

```{r}
plot(pbmc3k_nonnegL_pe$fit_flash$F.pm[,2],f0_init,xlab='f0 estimates',ylab='f0 init value',pch=20,col='grey70')
abline(a=0,b=1)
```

```{r}
plot.factors(pbmc3k_nonnegL_pe$fit_flash,pbmc_facs$samples$subpop,nonnegative = T)
```

```{r}
structure_plot_general(pbmc3k_nonnegL_pe$fit_flash$L.pm,pbmc3k_nonnegL_pe$fit_flash$F.pm,pbmc_facs$samples$subpop,LD=T)
```


## Nonnegative loadings with unimodal nonnegative prior, and sparse factors

```{r}
pbmc3k_nonnegL_un = readRDS('output/pbmc_fasttopics/pbmc3k_nonnegL_un.rds')
```

```{r}
summary_plot(pbmc3k_nonnegL_un)
```

```{r}
plot(pbmc3k_nonnegL_un$fit_flash$F.pm[,2],f0_init,xlab='f0 estimates',ylab='f0 init value',pch=20,col='grey70')
abline(a=0,b=1)
```

```{r}
plot.factors(pbmc3k_nonnegL_un$fit_flash,pbmc_facs$samples$subpop,nonnegative = T)
```

```{r}
structure_plot_general(pbmc3k_nonnegL_un$fit_flash$L.pm,pbmc3k_nonnegL_un$fit_flash$F.pm,pbmc_facs$samples$subpop,LD=T)
```

## Nonnegative loadings and factors with point exponential prior

```{r}
pbmc3k_nonnegLF_pe = readRDS('output/pbmc_fasttopics/pbmc3k_nonnegLF_pe.rds')
```

```{r}
summary_plot(pbmc3k_nonnegLF_pe)
```

```{r}
plot(pbmc3k_nonnegLF_pe$fit_flash$F.pm[,2],f0_init,xlab='f0 estimates',ylab='f0 init value',pch=20,col='grey70')
abline(a=0,b=1)
```

```{r}
plot.factors(pbmc3k_nonnegLF_pe$fit_flash,pbmc_facs$samples$subpop,nonnegative = T)
```

```{r}
structure_plot_general(pbmc3k_nonnegLF_pe$fit_flash$L.pm,pbmc3k_nonnegLF_pe$fit_flash$F.pm,pbmc_facs$samples$subpop,LD=T)
```


## Nonnegative loadings and factors with unimodal nonnegative prior

```{r}
pbmc3k_nonnegLF_un = readRDS('output/pbmc_fasttopics/pbmc3k_nonnegLF_un.rds')
```

```{r}
summary_plot(pbmc3k_nonnegLF_un)
```


```{r}
plot(pbmc3k_nonnegLF_un$fit_flash$F.pm[,2],f0_init,xlab='f0 estimates',ylab='f0 init value',pch=20,col='grey70')
abline(a=0,b=1)
```

```{r}
plot.factors(pbmc3k_nonnegLF_un$fit_flash,pbmc_facs$samples$subpop,nonnegative = T)
```

```{r}
structure_plot_general(pbmc3k_nonnegLF_un$fit_flash$L.pm,pbmc3k_nonnegLF_un$fit_flash$F.pm,pbmc_facs$samples$subpop,LD=T)
```

